### ISSUE TITLE
Tf.gradients returning all zeroes when called on result of a tf.gradient call

### ISSUE THREAD

**Issue by rkjones4** [tensorflow/tensorflow#9398]
I have the following code snippet:

interpolates = alpha*real_data + ((1-alpha)*fake_data)
disc_interpolates = Discriminator(interpolates)
gradients = tf.gradients(disc_interpolates, [interpolates])[0]
second_grad = tf.gradients(gradients[0], [interpolates])[0]

Where Discriminator corresponds to a neural network. The first call to tf.gradients is working correctly and I get back non-zero slope values in the gradients variable. However whenever I try to find the second derivative by applying tf.gradients to the gradients variable, my result is always a vector of zeroes. 

Is this expected behavior or should I be able to find the second derivatives of my neural net?

**Comment by drpngx** [comment]
Yes, that is correct. This is a linear function whose second derivative is zero.

**Comment by praysunday** [comment]
tnanks  my English is terrible,but Ilove AI so much.

**Comment by rkjones4** [comment]
Here is the code for the discriminator:

def Discriminator(inputs):
    output = tf.reshape(inputs, [-1, 3, 32, 32])

    output = lib.ops.conv2d.Conv2D('Discriminator.1', 3, DIM, 5, output, stride=2)
    output = LeakyReLU(output)

    output = lib.ops.conv2d.Conv2D('Discriminator.2', DIM, 2*DIM, 5, output, stride=2)
    output = LeakyReLU(output)

    output = lib.ops.conv2d.Conv2D('Discriminator.3', 2*DIM, 4*DIM, 5, output, stride=2)
    output = LeakyReLU(output)

    output = tf.reshape(output, [-1, 4*4*4*DIM])
    output = lib.ops.linear.Linear('Discriminator.Output', 4*4*4*DIM, 1, output)

    return tf.reshape(output, [-1])

I don't think the second derivative will be perfectly linear in all places, and I would be very surprised if that is the case because the first order derivates are not all exactly one and have variances greater than >.1

**Comment by drpngx** [comment]
Well if you are in the linear part, relu is linear and convolution is a linear combination of inputs, so you should get a zero gradient. Replace relu with, say, sigmoid and you should get something sigmoid-like.

**Comment by girving** [comment]
It's leaky relus, so zero gradients are unlikely.  However, this question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is likely not a bug or feature request. There is also a larger community that reads questions there. Thanks!
